// sanity-plugin-vimeo-field/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var sanity = require('sanity');
var jsxRuntime = require('react/jsx-runtime');
var ui = require('@sanity/ui');
var react = require('react');
var icons = require('@sanity/icons');

const DataFetcher = (props) => {
  const { accessToken, onSuccess, fields } = props;
  const [vimeoId, setVimeoId] = react.useState("");
  const [isFetching, setIsFetching] = react.useState(false);
  const [errorMsg, setErrorMsg] = react.useState("");
  let vimeoFields;
  let url = "https://api.vimeo.com/videos/".concat(vimeoId, "?fields=name,play,pictures,files");
  if (fields == null ? void 0 : fields.length) {
    vimeoFields = fields == null ? void 0 : fields.join(",");
    url += ",".concat(vimeoFields);
  }
  const handleChange = (event) => {
    setVimeoId(event.target.value);
  };
  const handleSubmit = async () => {
    setIsFetching(true);
    const options = {
      method: "GET",
      headers: {
        Authorization: "Bearer ".concat(accessToken)
      }
    };
    try {
      const response = await fetch(url, options);
      const data = await response.json();
      if (data == null ? void 0 : data.name) {
        setErrorMsg("");
        data.id = vimeoId;
        onSuccess(data);
      } else if (data == null ? void 0 : data.error) {
        setErrorMsg(data == null ? void 0 : data.error);
      }
      setIsFetching(false);
    } catch (error) {
      console.error("Error fetching data:", error);
      setErrorMsg("There was an error fetching data.");
      setIsFetching(false);
    }
  };
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { children: [
    !accessToken && /* @__PURE__ */ jsxRuntime.jsxs(ui.Inline, { space: [2], children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: 2, children: "No" }),
      /* @__PURE__ */ jsxRuntime.jsx(ui.Code, { size: 2, children: "SANITY_STUDIO_VIMEO_ACCESS_TOKEN" }),
      /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: 2, children: "found!" })
    ] }),
    accessToken && // Fetcher
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Inline, { space: [2], children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.TextInput,
        {
          onChange: handleChange,
          value: vimeoId,
          placeholder: "Vimeo ID",
          disabled: isFetching
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        ui.Button,
        {
          text: "Fetch",
          onClick: handleSubmit,
          disabled: vimeoId.trim() === "" || isFetching,
          icon: icons.SyncIcon,
          tone: "primary"
        }
      ),
      isFetching && /* @__PURE__ */ jsxRuntime.jsx(ui.Spinner, { muted: true })
    ] }),
    errorMsg && /* @__PURE__ */ jsxRuntime.jsxs("p", { style: { color: "red" }, children: [
      "Error: ",
      errorMsg
    ] })
  ] });
};

const VideoInput = (config, props) => {
  var _a, _b, _c, _d;
  const { fields } = (_a = props == null ? void 0 : props.schemaType) == null ? void 0 : _a.options;
  const { accessToken } = config;
  const { onChange, value } = props;
  const handleReset = () => {
    onChange(sanity.unset());
  };
  const setVimeoData = (data) => {
    onChange(data ? sanity.set(data) : sanity.unset());
  };
  const imgStyle = {
    width: "auto",
    height: "100px",
    borderRadius: "2px"
  };
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Card, { children: [
    !value && /* @__PURE__ */ jsxRuntime.jsx(DataFetcher, { accessToken, onSuccess: setVimeoData, fields }),
    /* @__PURE__ */ jsxRuntime.jsx("div", { children: value == null ? void 0 : value.error }),
    ((_c = (_b = value == null ? void 0 : value.pictures) == null ? void 0 : _b.sizes) == null ? void 0 : _c.length) && /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { space: 4, children: [
      /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { space: 3, children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: 1, weight: "semibold", children: "ID" }),
        /* @__PURE__ */ jsxRuntime.jsx(ui.TextInput, { fontSize: 2, padding: 3, readOnly: true, value: value == null ? void 0 : value.id })
      ] }),
      /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { space: 3, children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: 1, weight: "semibold", children: "Thumbnail" }),
        /* @__PURE__ */ jsxRuntime.jsx(
          "img",
          {
            src: (_d = value.pictures.sizes[0]) == null ? void 0 : _d.link,
            width: value.pictures.sizes[0].width,
            height: value.pictures.sizes[0].height,
            alt: "Vimeo",
            style: imgStyle
          }
        )
      ] }),
      /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { space: 3, children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: 1, weight: "semibold", children: "Title" }),
        /* @__PURE__ */ jsxRuntime.jsx(ui.TextInput, { fontSize: 2, padding: 3, readOnly: true, value: value == null ? void 0 : value.name })
      ] }),
      /* @__PURE__ */ jsxRuntime.jsx(ui.Inline, { space: [2], children: /* @__PURE__ */ jsxRuntime.jsx(
        ui.Button,
        {
          text: "Reset",
          icon: icons.TrashIcon,
          mode: "ghost",
          onClick: handleReset,
          type: "reset",
          tone: "critical"
        }
      ) })
    ] })
  ] });
};

const defaultConfig = {
  accessToken: ""
};
const vimeoField = sanity.definePlugin((userConfig) => {
  const config = { ...defaultConfig, ...userConfig };
  return {
    name: "sanity-plugin-vimeo-field",
    schema: {
      types: [
        sanity.defineType({
          title: "Vimeo Video",
          name: "vimeo",
          type: "object",
          components: {
            input: (props) => VideoInput(config, props)
          },
          fields: [
            {
              type: "object",
              name: "vimeoData",
              title: "Vimeo Data",
              readOnly: true,
              fields: [
                {
                  title: "Vimeo ID",
                  name: "id",
                  type: "string"
                }
              ]
            }
          ]
        })
      ]
    }
  };
});

exports.vimeoField = vimeoField;
