// sanity-plugin-vimeo-field/dist/index.esm.js
import { unset, set, definePlugin, defineType } from 'sanity';
import { jsxs, jsx } from 'react/jsx-runtime';
import { Stack, Inline, Text, Code, TextInput, Button, Spinner, Card } from '@sanity/ui';
import { useState } from 'react';
import { SyncIcon, TrashIcon } from '@sanity/icons';

const DataFetcher = (props) => {
  const { accessToken, onSuccess, fields } = props;
  const [vimeoId, setVimeoId] = useState("");
  const [isFetching, setIsFetching] = useState(false);
  const [errorMsg, setErrorMsg] = useState("");
  let vimeoFields;
  let url = "https://api.vimeo.com/videos/".concat(vimeoId, "?fields=name,play,pictures,files");
  if (fields == null ? void 0 : fields.length) {
    vimeoFields = fields == null ? void 0 : fields.join(",");
    url += ",".concat(vimeoFields);
  }
  const handleChange = (event) => {
    setVimeoId(event.target.value);
  };
  const handleSubmit = async () => {
    setIsFetching(true);
    const options = {
      method: "GET",
      headers: {
        Authorization: "Bearer ".concat(accessToken)
      }
    };
    try {
      const response = await fetch(url, options);
      const data = await response.json();
      if (data == null ? void 0 : data.name) {
        setErrorMsg("");
        data.id = vimeoId;
        onSuccess(data);
      } else if (data == null ? void 0 : data.error) {
        setErrorMsg(data == null ? void 0 : data.error);
      }
      setIsFetching(false);
    } catch (error) {
      console.error("Error fetching data:", error);
      setErrorMsg("There was an error fetching data.");
      setIsFetching(false);
    }
  };
  return /* @__PURE__ */ jsxs(Stack, { children: [
    !accessToken && /* @__PURE__ */ jsxs(Inline, { space: [2], children: [
      /* @__PURE__ */ jsx(Text, { size: 2, children: "No" }),
      /* @__PURE__ */ jsx(Code, { size: 2, children: "SANITY_STUDIO_VIMEO_ACCESS_TOKEN" }),
      /* @__PURE__ */ jsx(Text, { size: 2, children: "found!" })
    ] }),
    accessToken && // Fetcher
    /* @__PURE__ */ jsxs(Inline, { space: [2], children: [
      /* @__PURE__ */ jsx(
        TextInput,
        {
          onChange: handleChange,
          value: vimeoId,
          placeholder: "Vimeo ID",
          disabled: isFetching
        }
      ),
      /* @__PURE__ */ jsx(
        Button,
        {
          text: "Fetch",
          onClick: handleSubmit,
          disabled: vimeoId.trim() === "" || isFetching,
          icon: SyncIcon,
          tone: "primary"
        }
      ),
      isFetching && /* @__PURE__ */ jsx(Spinner, { muted: true })
    ] }),
    errorMsg && /* @__PURE__ */ jsxs("p", { style: { color: "red" }, children: [
      "Error: ",
      errorMsg
    ] })
  ] });
};

const VideoInput = (config, props) => {
  var _a, _b, _c, _d;
  const { fields } = (_a = props == null ? void 0 : props.schemaType) == null ? void 0 : _a.options;
  const { accessToken } = config;
  const { onChange, value } = props;
  const handleReset = () => {
    onChange(unset());
  };
  const setVimeoData = (data) => {
    onChange(data ? set(data) : unset());
  };
  const imgStyle = {
    width: "auto",
    height: "100px",
    borderRadius: "2px"
  };
  return /* @__PURE__ */ jsxs(Card, { children: [
    !value && /* @__PURE__ */ jsx(DataFetcher, { accessToken, onSuccess: setVimeoData, fields }),
    /* @__PURE__ */ jsx("div", { children: value == null ? void 0 : value.error }),
    ((_c = (_b = value == null ? void 0 : value.pictures) == null ? void 0 : _b.sizes) == null ? void 0 : _c.length) && /* @__PURE__ */ jsxs(Stack, { space: 4, children: [
      /* @__PURE__ */ jsxs(Stack, { space: 3, children: [
        /* @__PURE__ */ jsx(Text, { size: 1, weight: "semibold", children: "ID" }),
        /* @__PURE__ */ jsx(TextInput, { fontSize: 2, padding: 3, readOnly: true, value: value == null ? void 0 : value.id })
      ] }),
      /* @__PURE__ */ jsxs(Stack, { space: 3, children: [
        /* @__PURE__ */ jsx(Text, { size: 1, weight: "semibold", children: "Thumbnail" }),
        /* @__PURE__ */ jsx(
          "img",
          {
            src: (_d = value.pictures.sizes[0]) == null ? void 0 : _d.link,
            width: value.pictures.sizes[0].width,
            height: value.pictures.sizes[0].height,
            alt: "Vimeo",
            style: imgStyle
          }
        )
      ] }),
      /* @__PURE__ */ jsxs(Stack, { space: 3, children: [
        /* @__PURE__ */ jsx(Text, { size: 1, weight: "semibold", children: "Title" }),
        /* @__PURE__ */ jsx(TextInput, { fontSize: 2, padding: 3, readOnly: true, value: value == null ? void 0 : value.name })
      ] }),
      /* @__PURE__ */ jsx(Inline, { space: [2], children: /* @__PURE__ */ jsx(
        Button,
        {
          text: "Reset",
          icon: TrashIcon,
          mode: "ghost",
          onClick: handleReset,
          type: "reset",
          tone: "critical"
        }
      ) })
    ] })
  ] });
};

const defaultConfig = {
  accessToken: ""
};
const vimeoField = definePlugin((userConfig) => {
  const config = { ...defaultConfig, ...userConfig };
  return {
    name: "sanity-plugin-vimeo-field",
    schema: {
      types: [
        defineType({
          title: "Vimeo Video",
          name: "vimeo",
          type: "object",
          components: {
            input: (props) => VideoInput(config, props)
          },
          fields: [
            {
              type: "object",
              name: "vimeoData",
              title: "Vimeo Data",
              readOnly: true,
              fields: [
                {
                  title: "Vimeo ID",
                  name: "id",
                  type: "string"
                }
              ]
            }
          ]
        })
      ]
    }
  };
});

export { vimeoField };
